File: src/main/java/com/tkr/desk/LotSplitter.java
package com.tkr.desk;

import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.List;

/**
 * Lot splitter for round/odd lots with pro-rata and sequential policies.
 * Ported from legacy-cpp lot_splitter.cc.
 */
public final class LotSplitter {

    public enum LotSplitPolicy { PRO_RATA_BY_WEIGHT, SEQUENTIAL_FILL, ODD_LOT_FIRST, ROUND_LOT_PRESERVE }

    public static final class LotSplitRequest {
        public int parentLotId;
        public int totalQtyMilli;
        public int roundLotMilli = 1000;
        public LotSplitPolicy policy = LotSplitPolicy.PRO_RATA_BY_WEIGHT;
        public List<Integer> targetAccountIds = new ArrayList<>();
        public List<Integer> targetWeightsBp = new ArrayList<>();
    }

    public static final class LotSplitFragment {
        public int fragmentId;
        public int parentLotId;
        public int accountId;
        public int qtyMilli;
        public boolean isOddLot;
        public int sequence;
    }

    public static final class LotSplitterConfig {
        public int defaultRoundLotMilli = 1000;
        public boolean assignFragmentIds = true;
    }

    public static final class LotSplitterResult {
        public Status status = Status.OK;
        public List<LotSplitFragment> fragments = new ArrayList<>();
        public int remainderMilli;
        public int oddLotCount;
        public int roundLotCount;
    }

    private final LotSplitterConfig config;
    private int nextFragmentId = 1;

    public LotSplitter() {
        this(new LotSplitterConfig());
    }

    public LotSplitter(LotSplitterConfig config) {
        this.config = config != null ? config : new LotSplitterConfig();
    }

    public static int countRoundLots(int qtyMilli, int roundLotMilli) {
        if (roundLotMilli == 0) return 0;
        return qtyMilli / roundLotMilli;
    }

    public static int oddLotRemainder(int qtyMilli, int roundLotMilli) {
        if (roundLotMilli == 0) return qtyMilli;
        return qtyMilli % roundLotMilli;
    }

    public void assignFragmentIds(List<LotSplitFragment> fragments) {
        if (fragments == null) return;
        for (LotSplitFragment frag : fragments) frag.fragmentId = nextFragmentId++;
    }

    public LotSplitterResult splitProRata(LotSplitRequest request) {
        LotSplitterResult result = new LotSplitterResult();
        result.status = Status.OK;
        if (request.targetAccountIds.isEmpty()) {
            result.status = Status.BOUNDS_ERROR;
            return result;
        }
        int totalWeight = 0;
        for (int w : request.targetWeightsBp) totalWeight += w;
        if (totalWeight == 0) totalWeight = request.targetAccountIds.size();
        int allocated = 0;
        for (int i = 0; i < request.targetAccountIds.size(); i++) {
            int weight = i < request.targetWeightsBp.size() ? request.targetWeightsBp.get(i) : 1;
            LotSplitFragment frag = new LotSplitFragment();
            frag.parentLotId = request.parentLotId;
            frag.accountId = request.targetAccountIds.get(i);
            frag.qtyMilli = request.totalQtyMilli * weight / totalWeight;
            frag.isOddLot = frag.qtyMilli < request.roundLotMilli;
            frag.sequence = i;
            allocated += frag.qtyMilli;
            result.fragments.add(frag);
        }
        int remainder = request.totalQtyMilli - allocated;
        result.remainderMilli = remainder;
        if (remainder > 0 && !result.fragments.isEmpty()) {
            result.fragments.get(0).qtyMilli += remainder;
        }
        assignFragmentIds(result.fragments);
        for (LotSplitFragment frag : result.fragments) {
            if (frag.isOddLot) result.oddLotCount++;
            else result.roundLotCount++;
        }
        return result;
    }

    public LotSplitterResult splitSequential(LotSplitRequest request) {
        LotSplitterResult result = new LotSplitterResult();
        result.status = Status.OK;
        int remaining = request.totalQtyMilli;
        int perAccount = request.totalQtyMilli / Math.max(1, request.targetAccountIds.size());
        for (int i = 0; i < request.targetAccountIds.size(); i++) {
            LotSplitFragment frag = new LotSplitFragment();
            frag.parentLotId = request.parentLotId;
            frag.accountId = request.targetAccountIds.get(i);
            frag.sequence = i;
            frag.qtyMilli = remaining >= perAccount ? perAccount : remaining;
            remaining -= frag.qtyMilli;
            frag.isOddLot = frag.qtyMilli < request.roundLotMilli;
            result.fragments.add(frag);
        }
        if (remaining > 0 && !result.fragments.isEmpty()) {
            result.fragments.get(result.fragments.size() - 1).qtyMilli += remaining;
            result.remainderMilli = 0;
        }
        assignFragmentIds(result.fragments);
        return result;
    }

    public LotSplitterResult splitOddLotFirst(LotSplitRequest request) {
        LotSplitterResult result = new LotSplitterResult();
        int odd = oddLotRemainder(request.totalQtyMilli, request.roundLotMilli);
        int roundQty = request.totalQtyMilli - odd;
        if (odd > 0 && !request.targetAccountIds.isEmpty()) {
            LotSplitFragment oddFrag = new LotSplitFragment();
            oddFrag.parentLotId = request.parentLotId;
            oddFrag.accountId = request.targetAccountIds.get(0);
            oddFrag.qtyMilli = odd;
            oddFrag.isOddLot = true;
            oddFrag.sequence = 0;
            result.fragments.add(oddFrag);
            result.oddLotCount++;
        }
        LotSplitRequest roundReq = copyRequest(request);
        roundReq.totalQtyMilli = roundQty;
        LotSplitterResult roundResult = splitProRata(roundReq);
        for (LotSplitFragment frag : roundResult.fragments) {
            frag.sequence += 1;
            result.fragments.add(frag);
            if (frag.isOddLot) result.oddLotCount++;
            else result.roundLotCount++;
        }
        result.status = roundResult.status;
        assignFragmentIds(result.fragments);
        return result;
    }

    public LotSplitterResult splitRoundLotPreserve(LotSplitRequest request) {
        LotSplitterResult result = new LotSplitterResult();
        int roundLot = request.roundLotMilli != 0 ? request.roundLotMilli : config.defaultRoundLotMilli;
        int remaining = request.totalQtyMilli;
        for (int i = 0; i < request.targetAccountIds.size() && remaining > 0; i++) {
            LotSplitFragment frag = new LotSplitFragment();
            frag.parentLotId = request.parentLotId;
            frag.accountId = request.targetAccountIds.get(i);
            frag.sequence = i;
            int fullLots = countRoundLots(remaining, roundLot);
            if (fullLots > 0) {
                frag.qtyMilli = roundLot;
                remaining -= roundLot;
            } else {
                frag.qtyMilli = remaining;
                remaining = 0;
            }
            frag.isOddLot = frag.qtyMilli < roundLot;
            if (frag.isOddLot) result.oddLotCount++;
            else result.roundLotCount++;
            result.fragments.add(frag);
        }
        result.remainderMilli = remaining;
        result.status = Status.OK;
        assignFragmentIds(result.fragments);
        return result;
    }

    private static LotSplitRequest copyRequest(LotSplitRequest request) {
        LotSplitRequest copy = new LotSplitRequest();
        copy.parentLotId = request.parentLotId;
        copy.totalQtyMilli = request.totalQtyMilli;
        copy.roundLotMilli = request.roundLotMilli;
        copy.policy = request.policy;
        copy.targetAccountIds = new ArrayList<>(request.targetAccountIds);
        copy.targetWeightsBp = new ArrayList<>(request.targetWeightsBp);
        return copy;
    }

    public LotSplitterResult split(LotSplitRequest request) {
        return switch (request.policy) {
            case SEQUENTIAL_FILL -> splitSequential(request);
            case ODD_LOT_FIRST -> splitOddLotFirst(request);
            case ROUND_LOT_PRESERVE -> splitRoundLotPreserve(request);
            default -> splitProRata(request);
        };
    }

    public LotSplitterResult splitBatch(BatchWireFrame frame, LotSplitPolicy policy) {
        LotSplitterResult combined = new LotSplitterResult();
        combined.status = Status.OK;
        for (WireBatchRecord rec : frame.records) {
            LotSplitRequest req = new LotSplitRequest();
            req.parentLotId = rec.recordId;
            req.totalQtyMilli = rec.qtyMilli;
            req.roundLotMilli = config.defaultRoundLotMilli;
            req.policy = policy;
            req.targetAccountIds.add(rec.accountId);
            req.targetWeightsBp.add(10000);
            LotSplitterResult partial = split(req);
            if (partial.status != Status.OK) {
                combined.status = partial.status;
                return combined;
            }
            combined.fragments.addAll(partial.fragments);
            combined.oddLotCount += partial.oddLotCount;
            combined.roundLotCount += partial.roundLotCount;
        }
        return combined;
    }
}

File: legacy-cpp/src/desk/lot_splitter.cc
#include "tkr/desk/lot_splitter.h"

#include <algorithm>

namespace tkr {
namespace desk {

LotSplitter::LotSplitter(LotSplitterConfig config)
    : config_(config), next_fragment_id_(1) {}

std::uint32_t LotSplitter::CountRoundLots(std::uint32_t qty_milli,
                                          std::uint32_t round_lot_milli) {
  if (round_lot_milli == 0) {
    return 0;
  }
  return qty_milli / round_lot_milli;
}

std::uint32_t LotSplitter::OddLotRemainder(std::uint32_t qty_milli,
                                           std::uint32_t round_lot_milli) {
  if (round_lot_milli == 0) {
    return qty_milli;
  }
  return qty_milli % round_lot_milli;
}

void LotSplitter::AssignFragmentIds(std::vector<LotSplitFragment>* fragments) {
  if (fragments == nullptr) {
    return;
  }
  for (LotSplitFragment& frag : *fragments) {
    frag.fragment_id = next_fragment_id_++;
  }
}

LotSplitterResult LotSplitter::SplitProRata(const LotSplitRequest& request) {
  LotSplitterResult result{};
  result.status = Status::kOk;

  if (request.target_account_ids.empty()) {
    result.status = Status::kBoundsError;
    return result;
  }

  std::uint32_t total_weight = 0;
  for (std::uint32_t w : request.target_weights_bp) {
    total_weight += w;
  }
  if (total_weight == 0) {
    total_weight = static_cast<std::uint32_t>(request.target_account_ids.size());
  }

  std::uint32_t allocated = 0;
  for (std::size_t i = 0; i < request.target_account_ids.size(); ++i) {
    const std::uint32_t weight =
        (i < request.target_weights_bp.size()) ? request.target_weights_bp[i] : 1;
    LotSplitFragment frag{};
    frag.parent_lot_id = request.parent_lot_id;
    frag.account_id = request.target_account_ids[i];
    frag.qty_milli =
        (request.total_qty_milli * weight) / total_weight;
    frag.is_odd_lot =
        frag.qty_milli < request.round_lot_milli;
    frag.sequence = static_cast<std::uint32_t>(i);
    allocated += frag.qty_milli;
    result.fragments.push_back(frag);
  }

  std::uint32_t remainder = request.total_qty_milli - allocated;
  result.remainder_milli = remainder;
  if (remainder > 0 && !result.fragments.empty()) {
    result.fragments[0].qty_milli += remainder;
  }

  AssignFragmentIds(&result.fragments);
  for (const LotSplitFragment& frag : result.fragments) {
    if (frag.is_odd_lot) {
      ++result.odd_lot_count;
    } else {
      ++result.round_lot_count;
    }
  }

  return result;
}

LotSplitterResult LotSplitter::SplitSequential(const LotSplitRequest& request) {
  LotSplitterResult result{};
  result.status = Status::kOk;

  std::uint32_t remaining = request.total_qty_milli;
  for (std::size_t i = 0; i < request.target_account_ids.size(); ++i) {
    LotSplitFragment frag{};
    frag.parent_lot_id = request.parent_lot_id;
    frag.account_id = request.target_account_ids[i];
    frag.sequence = static_cast<std::uint32_t>(i);

    const std::uint32_t per_account =
        request.total_qty_milli /
        static_cast<std::uint32_t>(request.target_account_ids.size());
    frag.qty_milli = (remaining >= per_account) ? per_account : remaining;
    remaining -= frag.qty_milli;
    frag.is_odd_lot = frag.qty_milli < request.round_lot_milli;
    result.fragments.push_back(frag);
  }

  if (remaining > 0 && !result.fragments.empty()) {
    result.fragments.back().qty_milli += remaining;
    result.remainder_milli = 0;
  }

  AssignFragmentIds(&result.fragments);
  return result;
}

LotSplitterResult LotSplitter::SplitOddLotFirst(const LotSplitRequest& request) {
  LotSplitterResult result{};
  const std::uint32_t odd = OddLotRemainder(request.total_qty_milli,
                                            request.round_lot_milli);
  const std::uint32_t round_qty = request.total_qty_milli - odd;

  if (odd > 0 && !request.target_account_ids.empty()) {
    LotSplitFragment odd_frag{};
    odd_frag.parent_lot_id = request.parent_lot_id;
    odd_frag.account_id = request.target_account_ids[0];
    odd_frag.qty_milli = odd;
    odd_frag.is_odd_lot = true;
    odd_frag.sequence = 0;
    result.fragments.push_back(odd_frag);
    ++result.odd_lot_count;
  }

  LotSplitRequest round_req = request;
  round_req.total_qty_milli = round_qty;
  LotSplitterResult round_result = SplitProRata(round_req);
  for (LotSplitFragment& frag : round_result.fragments) {
    frag.sequence += 1;
    result.fragments.push_back(frag);
    if (frag.is_odd_lot) {
      ++result.odd_lot_count;
    } else {
      ++result.round_lot_count;
    }
  }

  result.status = round_result.status;
  AssignFragmentIds(&result.fragments);
  return result;
}

LotSplitterResult LotSplitter::SplitRoundLotPreserve(
    const LotSplitRequest& request) {
  LotSplitterResult result{};
  const std::uint32_t round_lot =
      request.round_lot_milli ? request.round_lot_milli
                              : config_.default_round_lot_milli;

  std::uint32_t remaining = request.total_qty_milli;
  for (std::size_t i = 0; i < request.target_account_ids.size() && remaining > 0;
       ++i) {
    LotSplitFragment frag{};
    frag.parent_lot_id = request.parent_lot_id;
    frag.account_id = request.target_account_ids[i];
    frag.sequence = static_cast<std::uint32_t>(i);

    const std::uint32_t full_lots =
        CountRoundLots(remaining, round_lot);
    if (full_lots > 0) {
      frag.qty_milli = round_lot;
      remaining -= round_lot;
    } else {
      frag.qty_milli = remaining;
      remaining = 0;
    }

    frag.is_odd_lot = frag.qty_milli < round_lot;
    if (frag.is_odd_lot) {
      ++result.odd_lot_count;
    } else {
      ++result.round_lot_count;
    }
    result.fragments.push_back(frag);
  }

  result.remainder_milli = remaining;
  result.status = Status::kOk;
  AssignFragmentIds(&result.fragments);
  return result;
}

LotSplitterResult LotSplitter::Split(const LotSplitRequest& request) {
  switch (request.policy) {
    case LotSplitPolicy::kSequentialFill:
      return SplitSequential(request);
    case LotSplitPolicy::kOddLotFirst:
      return SplitOddLotFirst(request);
    case LotSplitPolicy::kRoundLotPreserve:
      return SplitRoundLotPreserve(request);
    case LotSplitPolicy::kProRataByWeight:
    default:
      return SplitProRata(request);
  }
}

LotSplitterResult LotSplitter::SplitBatch(const BatchWireFrame& frame,
                                          LotSplitPolicy policy) {
  LotSplitterResult combined{};
  combined.status = Status::kOk;

  for (const WireBatchRecord& rec : frame.records) {
    LotSplitRequest req{};
    req.parent_lot_id = rec.record_id;
    req.total_qty_milli = rec.qty_milli;
    req.round_lot_milli = config_.default_round_lot_milli;
    req.policy = policy;
    req.target_account_ids.push_back(rec.account_id);
    req.target_weights_bp.push_back(10000);

    LotSplitterResult partial = Split(req);
    if (partial.status != Status::kOk) {
      combined.status = partial.status;
      return combined;
    }

    for (LotSplitFragment& frag : partial.fragments) {
      combined.fragments.push_back(frag);
    }
    combined.odd_lot_count += partial.odd_lot_count;
    combined.round_lot_count += partial.round_lot_count;
  }

  return combined;
}

}  // namespace desk
}  // namespace tkr
